### PR TITLE
remove config deserialize from core

### PR DESCRIPTION
**Pull request by kazuk**
## Issue number and link

Fixes: #220

## Describe your changes

- remove `config` crate dependency from `springql-core`
- add `config` crate to `springql`
- move config deserialize function from core, to `springql`
- add public surface api `config_from_toml`

## Checklist before requesting a review

- [x] I follow the [Semantic Pull Requests](https://github.com/zeke/semantic-pull-requests) rules (bugfix/feature)
- [x] I specified links to related issues (must: bugfix, want: feature)
- [x] I have performed a self-review of my code (bugfix/feature)
- [ ] I have added thorough tests (bugfix/feature)
- [x] I have edited `## [Unreleased]` section in `CHANGELOG.md` following [keep a changelog](https://keepachangelog.com/en/1.0.0/) syntax (bugfix/feature)
- [ ] I {made/will make} a related pull request for [documentation repo](https://github.com/SpringQL/SpringQL.github.io) (feature)

### CODE REVIEW

File: springql-core/src/api/spring_config.rs [hunk 1]
```diff
@@ -74,70 +13,51 @@ pub struct SpringConfig {
...
+            source_reader: SpringSourceReaderConfig {
+                net_connect_timeout_msec: 1000,
+                net_read_timeout_msec: 100,
+                can_read_timeout_msec: 100,
+            },
+            sink_writer: SpringSinkWriterConfig {
+                net_connect_timeout_msec: 1000,
+                net_write_timeout_msec: 100,
+            },
+        }
```
Comment (laysakura) [inline springql-core/src/api/spring_config.rs:46]: Don't want to hold default value both in https://github.com/SpringQL/SpringQL/pull/216/files#diff-56de30082a9c2e6459497fd78a8e2110be421f0674fa1c106ec2dbedfa952a71R12 and here.

How about changing `springql-core:: ... ::SpringConfig` into `spring-core:: ... ::Config`, which does not impl `Default`?

Then only `springql::SpringConfig` impl `Default` using `SPRING_CONFIG_DEFAULT`.

File: springql/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    ///   - `overwrite_config_toml` is not valid as TOML.
+    pub fn from_toml(overwrite_config_toml: &str) -> Result<SpringConfig> {
+        Self::load(overwrite_config_toml)
+    }
+}
+
+/// deserialize configration from toml
+pub fn config_from_toml(toml: &str) -> Result<SpringConfig> {
+    SpringConfigDeserialize::from_toml(toml)
+}
```
Comment (laysakura) [inline springql/src/config.rs:196]: I like this better:

- springql-core has `Config`, which does not impl Default and not visible to `springql` crate users.
- springql crate implements `SpringConfig`, which impl Default and visible to `springql` crate users. Also, it implements `pub fn new(overwrite_config_toml: &str) -> Result<Self>` as before.

- For what reason is the new API required?
- Why is this a free function (not inside `impl

**Review by laysakura — CHANGES_REQUESTED** [review]

**Comment by laysakura** [comment]
I guess we can keep `SpringConfig` API for `springql` crate users.
I feel It is totally OK to break the `springql-core`'s config related APIs but I would like to keep the API in `springql` crate.

File: springql-core/src/api/spring_config.rs (hunk 1 above)
Comment (kazuk) [reply springql-core/src/api/spring_config.rs:46]: this Default implementation used by only tests.

I move into `test_support`

https://github.com/SpringQL/SpringQL/pull/216/files#diff-7af0b986cf0abbb68962c12dbea4a8d1cd69bc2be040c92fe58c7a412848c61a

File: springql/src/config.rs (hunk 2 above)
Comment (kazuk) [reply springql/src/config.rs:196]: rust can not `impl` outside of crate.

```
   Compiling springql v0.15.0 (/home/kazuk/SpringQL/springql)
error[E0116]: cannot define inherent `impl` for a type outside of the crate where the type is defined
   --> springql/src/config.rs:193:1
    |
193 | / impl SpringConfig {
194 | |     pub fn new(toml: &str) -> Result<SpringConfig> {
195 | |         SpringConfigDeserialize::from_toml(toml)
196 | |     }
197 | | }
    | |_^ impl for type defined outside of crate.
    |
    = note: define and implement a trait or new type instead

For more information about this error, try `rustc --explain E0116`.
```

File: springql-core/src/api/spring_config.rs (hunk 1 above)
Comment (laysakura) [reply springql-core/src/api/spring_config.rs:46]: No. It's used in online documents and some users (at least me).

https://springql.github.io/get-started/install-and-run-app#create-a-bin-crate-to-write-a-demo
image

File: springql/src/config.rs (hunk 2 above)
Comment (laysakura) [reply springql/src/config.rs:196]: I know it. So I claim that we move `SpringConfig` from springql-core into `springql` crate and then impl Default there.

File: springql-core/src/api/spring_config.rs (hunk 1 above)
Comment (laysakura) [reply springql-core/src/api/spring_config.rs:46]: Ah, sorry. I might mistakenly get what you say.

I realized that:

- Default values in TOML is (maybe) a good idea.
- TOML will be moved to springql crate.
- springql-core needs some way to provide Default impl for tests inside the crate.

Umm...

File: springql/src/config.rs (hunk 2 above)
Comment (kazuk) [reply springql/src/config.rs:196]: turn `SpringConfig` definition `struct` to` trait`?

**Comment by kazuk** [comment]
I close this.

and Re-create pr to implement this.
https://github.com/SpringQL/SpringQL/pull/216#discussion_r919574717

**Pull request closed without merging**